crates/transpiler/src/passes/vf2/vf2_layout.rs

```diff
@@ -654,83 +816,90 @@ pub fn vf2_layout_pass(
-)]
-pub fn score_layout(
-    bit_list: PyReadonlyArray1<i32>,
-    edge_list: &EdgeList,
-    error_map: &ErrorMap,
-    layout: &NLayout,
-    strict_direction: bool,
-    run_in_parallel: bool,
-) -> PyResult<f64> {
-    let bit_counts = bit_list.as_slice()?;
-    let edge_filter_map = |(index_arr, gate_count): &([VirtualQubit; 2], i32)| -> Option<f64> {
-        let mut error = error_map
-            .error_map
-            .get(&[index_arr[0].to_phys(layout), index_arr[1].to_phys(layout)]);
-        if !strict_direction && error.is_none() {
-            error = error_map
-                .error_map
-                .get(&[index_arr[1].to_phys(layout), index_arr[0].to_phys(layout)]);
-        }
-        error.map(|error| {
-            if !error.is_nan() {
-                (1. - error).powi(*gate_count)
+#[pyo3(signature = (dag, target, config))]
+pub fn vf2_layout_pass_exact(
+    dag: &DAGCircuit,
+    target: &Target,
+    config: &Vf2PassConfiguration,
+) -> PyResult<Vf2PassReturn> {
+    let Some((mut coupling_graph, interner)) = build_exact_coupling_map(target) else {
+        return Ok(Vf2PassReturn::NoSolution);
+    };
+    let add_interaction =
+        |uses: &mut Vec<(Interned<str>, usize)>, inst: &PackedInstruction, repeats: usize| {
+            let Some(key) = interner.try_key(inst.op.name()) else {
+                return false;
+            };
+            if let Some((_, count)) = uses.iter_mut().find(|(name, _)| key == *name) {
+                *count += repeats;
             } else {
-                1.
+                uses.push((key, repeats));
             }
-        })
+            true
+        };
+    let Some(mut interactions) = VirtualInteractions::from_dag(dag, add_interaction)? else {
+        return Ok(Vf2PassReturn::NoSolution);
     };
-    let bit_filter_map = |(v_bit_index, gate_counts): (usize, &i32)| -> Option<f64> {
-        let p_bit = VirtualQubit::new(v_bit_index.try_into().unwrap()).to_phys(layout);
-        let error = error_map.error_map.get(&[p_bit, p_bit]);
 
-        error.map(|error| {
-            if !error.is_nan() {
-                (1. - error).powi(*gate_counts)
-            } else {
-                1.
-            }
-        })
+    // The optimisation we have in the "average" case where we assign loose 1q gates after matching
+    // the rest of the graph doesn't hold in the "exact" case, so we have to have VF2 match them all
+    // at once.  For example, consider a heterogeneous target where only one qubit has `rx`
+    // available, but VF2 matches a non-`rx`-using qubit onto that one during the first step,
+    // because it isn't aware there's a necessary` rx` gate in the free list.
+    if interactions.uncoupled.len() > 1 {
+        // ... but if there's too many uncoupled qubits, there's a combinatorial explosion in the
+        // matching time, so we just bail out.  It would be nice to deal with this better.
+        return Ok(Vf2PassReturn::NoSolution);
+    }
```

Comment:
Do we have this problem even after the new early-pruning mechanism? Hmm, yeah, I guess it is. However, I am still wondering if it makes more sense to eagerly give up rather than performing the allowed `call_limit` number of calls. After all, it is also possible that we will find a better solution.

Reply:
the initial limit in `call_limit` is typically set rather high, so it can totally blow out the time of the algorithm with no real benefit. This would be nice to revisit, but it's possibly better as follow-on research?

Reply:
Agreed, it's probably best not to change the default parameters right now. However, it does feel strange to apply such "hacks" when there is a very natural `call_limit` option that is specifically there to bound the running time of the algorithm.